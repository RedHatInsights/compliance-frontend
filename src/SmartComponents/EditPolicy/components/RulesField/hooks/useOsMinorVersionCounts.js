import { useState, useEffect } from 'react';
import { mapCountOsMinorVersions } from 'Store/Reducers/SystemStore';

const profilesToOsMinorMap = (profiles, hosts) =>
  (profiles || []).reduce((acc, profile) => {
    if (profile.osMinorVersion !== '') {
      acc[profile.osMinorVersion] ||= {
        osMinorVersion: profile.osMinorVersion,
        count: 0,
      };
    }

    return acc;
  }, mapCountOsMinorVersions(hosts || []));

const policyProfilesMinorVersions = (policy) =>
  policy?.policy?.profiles
    .map(({ osMinorVersion }) => osMinorVersion)
    .filter((version) => version.length > 0);

const useOsMinorVersionCounts = (policy, systems) => {
  const [osMinorVersionCounts, setOsMinorVersionCounts] = useState({
    hasNewOS: false,
    policyOsMajorVersion: policy?.osMajorVersion,
    policyOsMinorVersions: [],
    systemsOsMinorVersions: [],
    systemsOsMinorVersionCounts: {},
  });

  useEffect(() => {
    const systemsOsMinorVersions = Object.keys(
      mapCountOsMinorVersions(systems?.value || [])
    );
    const policyOsMinorVersions = policyProfilesMinorVersions(policy);

    setOsMinorVersionCounts((currentState) => ({
      ...currentState,
      policyOsMinorVersions,
      systemsOsMinorVersions,
      hasNewOS: systemsOsMinorVersions > policyOsMinorVersions,
      systemsOsMinorVersionCounts: profilesToOsMinorMap(
        policy?.policy?.profiles,
        systems?.value
      ),
    }));
  }, [policy, systems]);

  return osMinorVersionCounts;
};

export default useOsMinorVersionCounts;
