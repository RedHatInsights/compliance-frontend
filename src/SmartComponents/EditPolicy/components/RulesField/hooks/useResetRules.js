import { useCallback, useEffect } from 'react';
import useFormApi from '@data-driven-forms/react-form-renderer/use-form-api';
import { profilesWithRulesToSelection } from 'PresentationalComponents/TabbedRules';
import {
  getBenchmarkBySupportedOsMinor,
  getBenchmarkProfile,
} from '../helpers';

const useResetRules = (policy, systemsOsMinorVersions, benchmarks) => {
  const { change, getFieldState } = useFormApi();
  const rules = getFieldState('rules');
  // Default rules for each profile/benchmark
  // const defaultRules = useMemo(() => {}, []);

  // Rules from the policy set by the user
  // const policyRules = useMemo(() => {}, []);

  const reset = useCallback(() => {
    change('rules');
  }, [change, rules]);

  useEffect(() => {
    const policyProfilesMinorVersions = policy?.policy?.profiles
      .map(({ osMinorVersion }) => osMinorVersion)
      .filter((version) => version.length > 0);
    const newSystemsOsMinorVersions = systemsOsMinorVersions.filter(
      (version) => !policyProfilesMinorVersions.includes(version)
    );
    const policyHasNewMinorVersions = newSystemsOsMinorVersions.length > 0;

    if (policyHasNewMinorVersions) {
      console.log('Has new versions', benchmarks, rules.value);
      const currentProfileIds = rules.value.map(({ id }) => id);
      const newProfiles = newSystemsOsMinorVersions
        .map((version) => {
          const benchmarkProfile = getBenchmarkProfile(
            getBenchmarkBySupportedOsMinor(benchmarks || [], version),
            policy.refId
          );

          return benchmarkProfile
            ? {
                ...benchmarkProfile,
                osMinorVersion: version,
              }
            : undefined;
        })
        .filter(
          (profile) => !!profile && !currentProfileIds.includes(profile.id)
        );
      console.log(newProfiles);
      if (newProfiles.length > 0) {
        const newRules = profilesWithRulesToSelection(newProfiles);
        console.log(rules.value, newRules);
        change('rules', [...rules.value, ...newRules]);
      }
    }
  }, [policy, systemsOsMinorVersions, benchmarks, rules]);

  return {
    reset,
  };
};

export default useResetRules;
