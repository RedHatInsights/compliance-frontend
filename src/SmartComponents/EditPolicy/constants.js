import gql from 'graphql-tag';

export const PROFILES_QUERY = gql`
  query Profiles($filter: String!) {
    profiles(search: $filter) {
      edges {
        node {
          id
          name
          refId
          osMinorVersion
          osMajorVersion
          policy {
            id
          }
          policyType
          benchmark {
            id
            refId
            latestSupportedOsMinorVersions
            osMajorVersion
            version
          }
          rules {
            id
            title
            severity
            rationale
            refId
            description
            remediationAvailable
            identifier
          }
        }
      }
    }
  }
`;

export const BENCHMARKS_QUERY = gql`
  query Benchmarks($filter: String!) {
    benchmarks(search: $filter) {
      nodes {
        id
        latestSupportedOsMinorVersions
        profiles {
          id
          refId
          ssgVersion
        }
      }
    }
  }
`;
