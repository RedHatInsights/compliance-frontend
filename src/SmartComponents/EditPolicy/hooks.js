import { useCallback, useState } from 'react';
import { usePolicy } from 'Mutations';
import { useLinkToBackground, useAnchor } from 'Utilities/Router';
import { dispatchNotification } from 'Utilities/Dispatcher';

export const useLinkToPolicy = () => {
  const anchor = useAnchor();
  const linkToBackground = useLinkToBackground('/scappolicies');
  return () => {
    linkToBackground({ hash: anchor });
  };
};

export const useOnSave = (policy, updatedPolicyHostsAndRules) => {
  const updatePolicy = usePolicy();
  const linkToPolicy = useLinkToPolicy();
  const [isSaving, setIsSaving] = useState(false);

  const onSave = useCallback(() => {
    if (isSaving) {
      return Promise.resolve({});
    }

    setIsSaving(true);
    updatePolicy(policy, updatedPolicyHostsAndRules)
      .then(() => {
        setIsSaving(false);
        dispatchNotification({
          variant: 'success',
          title: 'Policy updated',
          autoDismiss: true,
        });
        linkToPolicy();
      })
      .catch((error) => {
        setIsSaving(false);
        dispatchNotification({
          variant: 'danger',
          title: 'Error updating policy',
          description: error.message,
        });
        linkToPolicy();
      });
  }, [isSaving, policy, updatedPolicyHostsAndRules]);

  return [isSaving, onSave];
};
export const useSavePolicyDetails = (policyId) => {
  const anchor = useAnchor();
  const linkToBackground = useLinkToBackground(`/scappolicies/${policyId}`);
  return () => {
    linkToBackground({ hash: anchor });
  };
};

export const useOnSavePolicyDetails = (
  policy,
  updatedPolicyHostsAndRules,
  closingFunction,
  policyId
) => {
  const updatePolicy = usePolicy();
  const savePolicyDetails = useSavePolicyDetails(policyId);
  const [isSaving, setIsSaving] = useState(false);
  const onSave = () => {
    setIsSaving(true);
    closingFunction();
    updatePolicy(policy, updatedPolicyHostsAndRules)
      .then(() => {
        setIsSaving(false);
        dispatchNotification({
          variant: 'success',
          title: 'Policy updated',
          autoDismiss: true,
        });
        savePolicyDetails();
      })
      .catch((error) => {
        setIsSaving(false);
        dispatchNotification({
          variant: 'danger',
          title: 'Error updating policy',
          description: error.message,
        });
        savePolicyDetails();
      });
  };
  return [isSaving, onSave];
};
