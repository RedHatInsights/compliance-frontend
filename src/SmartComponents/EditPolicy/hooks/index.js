import { useState } from 'react';

export const useNewRulesAlertState = (...args) => useState(...args);
