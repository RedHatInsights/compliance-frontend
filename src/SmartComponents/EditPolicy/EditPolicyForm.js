import React, { useCallback, useState, useEffect } from 'react';
import propTypes from 'prop-types';
import { Form, Tab, TabTitleText } from '@patternfly/react-core';
import { RoutedTabs } from 'PresentationalComponents';
import EditPolicyRulesTab from './EditPolicyRulesTab';
import EditPolicySystemsTab from './EditPolicySystemsTab';
import NewRulesAlert from './components/NewRulesAlert';
import { mapCountOsMinorVersions } from 'Store/Reducers/SystemStore';
import { profilesWithRulesToSelection } from 'PresentationalComponents/TabbedRules';
import { thresholdValid } from '../CreatePolicy/validate';
import { useNewRulesAlertState } from './hooks/index';

const profilesToOsMinorMap = (profiles, hosts) =>
  (profiles || []).reduce((acc, profile) => {
    if (profile.osMinorVersion !== '') {
      acc[profile.osMinorVersion] ||= {
        osMinorVersion: profile.osMinorVersion,
        count: 0,
      };
    }

    return acc;
  }, mapCountOsMinorVersions(hosts || []));

const EditPolicyForm = ({
  policy,
  setUpdatedPolicy,
  selectedRuleRefIds,
  setSelectedRuleRefIds,
  selectedSystems,
  setSelectedSystems,
  setRuleValues,
  ruleValues,
}) => {
  const policyProfiles = policy?.policy?.profiles || [];
  const [osMinorVersionCounts, setOsMinorVersionCounts] = useState({});
  const [newRulesAlert, setNewRulesAlert] = useNewRulesAlertState(false);

  const handleSystemSelect = useCallback(
    (newSelectedSystems) => {
      const policyMinorVersions = policy.hosts.map(
        ({ osMinorVersion }) => osMinorVersion
      );
      const hasNewOsMinorVersions =
        newSelectedSystems.filter(
          ({ osMinorVersion }) => !policyMinorVersions.includes(osMinorVersion)
        ).length > 0;

      setSelectedSystems(newSelectedSystems);
      setNewRulesAlert(hasNewOsMinorVersions);
      setOsMinorVersionCounts(
        profilesToOsMinorMap(policyProfiles, newSelectedSystems)
      );
    },
    [policyProfiles, selectedRuleRefIds]
  );

  useEffect(() => {
    if (policy) {
      const complianceThresholdValid = thresholdValid(
        policy.complianceThreshold
      );
      const profilesWithOsMinor = policyProfiles.filter(
        ({ osMinorVersion }) => !!osMinorVersion
      );
      setUpdatedPolicy({
        ...policy,
        complianceThresholdValid,
      });

      setSelectedRuleRefIds(profilesWithRulesToSelection(profilesWithOsMinor));
      handleSystemSelect(policy.hosts);
    }
  }, [policy]);

  return (
    <Form>
      <RoutedTabs ouiaId="EditSystems" defaultTab="systems" id="policy-tabs">
        <Tab
          eventKey="rules"
          ouiaId="Rules"
          title={<TabTitleText>Rules</TabTitleText>}
        >
          <EditPolicyRulesTab
            policy={policy}
            setSelectedRuleRefIds={setSelectedRuleRefIds}
            setRuleValues={setRuleValues}
            ruleValues={ruleValues}
            selectedRuleRefIds={selectedRuleRefIds}
            osMinorVersionCounts={osMinorVersionCounts}
          />
        </Tab>
        <Tab
          eventKey="systems"
          ouiaId="Systems"
          title={<TabTitleText>Systems</TabTitleText>}
        >
          <EditPolicySystemsTab
            policy={policy}
            selectedSystems={selectedSystems}
            onSystemSelect={handleSystemSelect}
          />
          {newRulesAlert && <NewRulesAlert />}
        </Tab>
      </RoutedTabs>
    </Form>
  );
};

EditPolicyForm.propTypes = {
  policy: propTypes.object,
  updatedPolicy: propTypes.object,
  setUpdatedPolicy: propTypes.func,
  selectedRuleRefIds: propTypes.arrayOf(propTypes.object),
  setSelectedRuleRefIds: propTypes.func,
  setSelectedSystems: propTypes.func,
  selectedSystems: propTypes.array,
  setRuleValues: propTypes.func,
  ruleValues: propTypes.array,
};

export default EditPolicyForm;
