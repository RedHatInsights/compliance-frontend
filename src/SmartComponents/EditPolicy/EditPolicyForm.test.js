import propTypes from 'prop-types';
import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { BENCHMARKS_QUERY } from './constants';

import { policies } from '@/__fixtures__/policies.js';

import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { Provider } from 'react-redux';
import { createStore } from 'redux';

import EditPolicyForm from './EditPolicyForm';
import { useNewRulesAlertState } from './hooks';

jest.mock('./hooks', () => ({
  ...jest.requireActual('./hooks'),
  useNewRulesAlertState: jest.fn(() => [false, () => false]),
}));

const initialState = {};
const reducer = (state = initialState, action) => {
  switch (action.type) {
    default:
      return state;
  }
};

const TestWrapper = ({ children, mocks = [] }) => {
  const history = createMemoryHistory();
  const store = createStore(reducer, initialState);

  return (
    <Provider store={store}>
      <Router history={history}>
        <MockedProvider mocks={mocks}>{children}</MockedProvider>
      </Router>
    </Provider>
  );
};

TestWrapper.propTypes = {
  children: propTypes.node,
  mocks: propTypes.arra,
};

describe('EditPolicyForm', () => {
  const policy = {
    ...policies.edges[0].node,
    supportedOsVersions: ['7.8', '7.9'],
  };
  const mocks = [
    {
      request: {
        query: BENCHMARKS_QUERY,
      },
      result: {
        data: {},
      },
    },
  ];
  const defaultProps = {
    setUpdatedPolicy: () => ({}),
    setSelectedRuleRefIds: () => ({}),
    setSelectedSystems: () => ({}),
    policy,
  };

  it('expect to render without error', () => {
    useNewRulesAlertState.mockImplementation(() => [false, () => false]);
    const { asFragment } = render(
      <TestWrapper mocks={mocks}>
        <EditPolicyForm {...defaultProps} />
      </TestWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('expect to render with alter', () => {
    useNewRulesAlertState.mockImplementation(() => [true, () => true]);
    const { asFragment } = render(
      <TestWrapper mocks={mocks}>
        <EditPolicyForm {...defaultProps} />
      </TestWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
