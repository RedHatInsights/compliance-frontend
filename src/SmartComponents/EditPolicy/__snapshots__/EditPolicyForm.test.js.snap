// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPolicyForm expect to render with alter 1`] = `
<DocumentFragment>
  <form
    class="pf-c-form"
    novalidate=""
  >
    <div
      class="pf-c-tabs"
      data-ouia-component-id="EditPolicy"
      data-ouia-component-type="PF4/Tabs"
      data-ouia-safe="true"
      id="policy-tabs"
    >
      <button
        aria-hidden="true"
        aria-label="Scroll left"
        class="pf-c-tabs__scroll-button"
        disabled=""
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 256 512"
          width="1em"
        >
          <path
            d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          />
        </svg>
      </button>
      <ul
        class="pf-c-tabs__list"
        role="tablist"
      >
        <li
          class="pf-c-tabs__item pf-m-current"
          role="presentation"
        >
          <button
            aria-controls="pf-tab-section-details-policy-tabs"
            aria-selected="true"
            class="pf-c-tabs__link"
            data-ouia-component-id="Details"
            data-ouia-component-type="PF4/TabButton"
            data-ouia-safe="true"
            id="pf-tab-details-policy-tabs"
            role="tab"
          >
            <span
              class="pf-c-tabs__item-text"
            >
              Details
            </span>
          </button>
        </li>
        <li
          class="pf-c-tabs__item"
          role="presentation"
        >
          <button
            aria-controls="pf-tab-section-rules-policy-tabs"
            aria-selected="false"
            class="pf-c-tabs__link"
            data-ouia-component-id="Rules"
            data-ouia-component-type="PF4/TabButton"
            data-ouia-safe="true"
            id="pf-tab-rules-policy-tabs"
            role="tab"
          >
            <span
              class="pf-c-tabs__item-text"
            >
              Rules
            </span>
          </button>
        </li>
        <li
          class="pf-c-tabs__item"
          role="presentation"
        >
          <button
            aria-controls="pf-tab-section-systems-policy-tabs"
            aria-selected="false"
            class="pf-c-tabs__link"
            data-ouia-component-id="Systems"
            data-ouia-component-type="PF4/TabButton"
            data-ouia-safe="true"
            id="pf-tab-systems-policy-tabs"
            role="tab"
          >
            <span
              class="pf-c-tabs__item-text"
            >
              Systems
            </span>
          </button>
        </li>
      </ul>
      <button
        aria-hidden="true"
        aria-label="Scroll right"
        class="pf-c-tabs__scroll-button"
        disabled=""
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 256 512"
          width="1em"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          />
        </svg>
      </button>
    </div>
    <section
      aria-labelledby="pf-tab-details-policy-tabs"
      class="pf-c-tab-content"
      data-ouia-component-id="Details"
      data-ouia-component-type="PF4/TabContent"
      data-ouia-safe="true"
      id="pf-tab-section-details-policy-tabs"
      role="tabpanel"
      tabindex="0"
    >
      <div
        class="pf-c-form"
      >
        <div
          class="pf-c-form__group"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="description"
            >
              <span
                class="pf-c-form__label-text"
              >
                Policy description
              </span>
              <span
                aria-hidden="true"
                class="pf-c-form__label-required"
              >
                 *
              </span>
            </label>
             
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <textarea
              aria-describedby="description"
              aria-invalid="false"
              class="pf-c-form-control"
              id="description"
              name="description"
              required=""
              style="width: 800px; height: 110px;"
              type="text"
            />
          </div>
        </div>
        <div
          class="pf-c-form__group"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="business-objective"
            >
              <span
                class="pf-c-form__label-text"
              >
                Business objective
              </span>
            </label>
             
            <span>
               
              <svg
                aria-hidden="true"
                class="grey-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </span>
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <input
              aria-describedby="business-objective"
              aria-invalid="false"
              class="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-3"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe="true"
              id="business-objective"
              name="business-objective"
              style="width: 300px;"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="pf-c-form__group"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="policy-threshold"
            >
              <span
                class="pf-c-form__label-text"
              >
                Compliance threshold (%)
              </span>
            </label>
             
            <span>
               
              <svg
                aria-hidden="true"
                class="grey-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </span>
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <input
              aria-describedby="policy-threshold"
              aria-invalid="false"
              class="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-4"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe="true"
              id="compliance-threshold"
              name="compliance-threshold"
              style="width: 150px;"
              type="number"
              value="100"
            />
            <div
              aria-live="polite"
              class="pf-c-form__helper-text"
              id="policy-threshold-helper"
            >
              A value of 95% or higher is recommended
            </div>
          </div>
        </div>
      </div>
    </section>
    <section
      aria-labelledby="pf-tab-rules-policy-tabs"
      class="pf-c-tab-content"
      data-ouia-component-id="Rules"
      data-ouia-component-type="PF4/TabContent"
      data-ouia-safe="true"
      hidden=""
      id="pf-tab-section-rules-policy-tabs"
      role="tabpanel"
      tabindex="0"
    >
      <div
        class="ins-c-table__empty"
      >
         
        <div
          class="ins-c-spinner"
          role="status"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading...
          </span>
        </div>
      </div>
    </section>
    <section
      aria-labelledby="pf-tab-systems-policy-tabs"
      class="pf-c-tab-content"
      data-ouia-component-id="Systems"
      data-ouia-component-type="PF4/TabContent"
      data-ouia-safe="true"
      hidden=""
      id="pf-tab-section-systems-policy-tabs"
      role="tabpanel"
      tabindex="0"
    >
      <section
        class="inventory"
      >
        <div>
          <h1>
            Inventory mock
          </h1>
        </div>
      </section>
    </section>
  </form>
</DocumentFragment>
`;

exports[`EditPolicyForm expect to render without error 1`] = `
<DocumentFragment>
  <form
    class="pf-c-form"
    novalidate=""
  >
    <div
      class="pf-c-tabs"
      data-ouia-component-id="EditPolicy"
      data-ouia-component-type="PF4/Tabs"
      data-ouia-safe="true"
      id="policy-tabs"
    >
      <button
        aria-hidden="true"
        aria-label="Scroll left"
        class="pf-c-tabs__scroll-button"
        disabled=""
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 256 512"
          width="1em"
        >
          <path
            d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          />
        </svg>
      </button>
      <ul
        class="pf-c-tabs__list"
        role="tablist"
      >
        <li
          class="pf-c-tabs__item pf-m-current"
          role="presentation"
        >
          <button
            aria-controls="pf-tab-section-details-policy-tabs"
            aria-selected="true"
            class="pf-c-tabs__link"
            data-ouia-component-id="Details"
            data-ouia-component-type="PF4/TabButton"
            data-ouia-safe="true"
            id="pf-tab-details-policy-tabs"
            role="tab"
          >
            <span
              class="pf-c-tabs__item-text"
            >
              Details
            </span>
          </button>
        </li>
        <li
          class="pf-c-tabs__item"
          role="presentation"
        >
          <button
            aria-controls="pf-tab-section-rules-policy-tabs"
            aria-selected="false"
            class="pf-c-tabs__link"
            data-ouia-component-id="Rules"
            data-ouia-component-type="PF4/TabButton"
            data-ouia-safe="true"
            id="pf-tab-rules-policy-tabs"
            role="tab"
          >
            <span
              class="pf-c-tabs__item-text"
            >
              Rules
            </span>
          </button>
        </li>
        <li
          class="pf-c-tabs__item"
          role="presentation"
        >
          <button
            aria-controls="pf-tab-section-systems-policy-tabs"
            aria-selected="false"
            class="pf-c-tabs__link"
            data-ouia-component-id="Systems"
            data-ouia-component-type="PF4/TabButton"
            data-ouia-safe="true"
            id="pf-tab-systems-policy-tabs"
            role="tab"
          >
            <span
              class="pf-c-tabs__item-text"
            >
              Systems
            </span>
          </button>
        </li>
      </ul>
      <button
        aria-hidden="true"
        aria-label="Scroll right"
        class="pf-c-tabs__scroll-button"
        disabled=""
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 256 512"
          width="1em"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          />
        </svg>
      </button>
    </div>
    <section
      aria-labelledby="pf-tab-details-policy-tabs"
      class="pf-c-tab-content"
      data-ouia-component-id="Details"
      data-ouia-component-type="PF4/TabContent"
      data-ouia-safe="true"
      id="pf-tab-section-details-policy-tabs"
      role="tabpanel"
      tabindex="0"
    >
      <div
        class="pf-c-form"
      >
        <div
          class="pf-c-form__group"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="description"
            >
              <span
                class="pf-c-form__label-text"
              >
                Policy description
              </span>
              <span
                aria-hidden="true"
                class="pf-c-form__label-required"
              >
                 *
              </span>
            </label>
             
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <textarea
              aria-describedby="description"
              aria-invalid="false"
              class="pf-c-form-control"
              id="description"
              name="description"
              required=""
              style="width: 800px; height: 110px;"
              type="text"
            />
          </div>
        </div>
        <div
          class="pf-c-form__group"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="business-objective"
            >
              <span
                class="pf-c-form__label-text"
              >
                Business objective
              </span>
            </label>
             
            <span>
               
              <svg
                aria-hidden="true"
                class="grey-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </span>
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <input
              aria-describedby="business-objective"
              aria-invalid="false"
              class="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-1"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe="true"
              id="business-objective"
              name="business-objective"
              style="width: 300px;"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="pf-c-form__group"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="policy-threshold"
            >
              <span
                class="pf-c-form__label-text"
              >
                Compliance threshold (%)
              </span>
            </label>
             
            <span>
               
              <svg
                aria-hidden="true"
                class="grey-icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                />
              </svg>
            </span>
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <input
              aria-describedby="policy-threshold"
              aria-invalid="false"
              class="pf-c-form-control"
              data-ouia-component-id="OUIA-Generated-TextInputBase-2"
              data-ouia-component-type="PF4/TextInput"
              data-ouia-safe="true"
              id="compliance-threshold"
              name="compliance-threshold"
              style="width: 150px;"
              type="number"
              value="100"
            />
            <div
              aria-live="polite"
              class="pf-c-form__helper-text"
              id="policy-threshold-helper"
            >
              A value of 95% or higher is recommended
            </div>
          </div>
        </div>
      </div>
    </section>
    <section
      aria-labelledby="pf-tab-rules-policy-tabs"
      class="pf-c-tab-content"
      data-ouia-component-id="Rules"
      data-ouia-component-type="PF4/TabContent"
      data-ouia-safe="true"
      hidden=""
      id="pf-tab-section-rules-policy-tabs"
      role="tabpanel"
      tabindex="0"
    >
      <div
        class="ins-c-table__empty"
      >
         
        <div
          class="ins-c-spinner"
          role="status"
        >
          <span
            class="pf-u-screen-reader"
          >
            Loading...
          </span>
        </div>
      </div>
    </section>
    <section
      aria-labelledby="pf-tab-systems-policy-tabs"
      class="pf-c-tab-content"
      data-ouia-component-id="Systems"
      data-ouia-component-type="PF4/TabContent"
      data-ouia-safe="true"
      hidden=""
      id="pf-tab-section-systems-policy-tabs"
      role="tabpanel"
      tabindex="0"
    >
      <section
        class="inventory"
      >
        <div>
          <h1>
            Inventory mock
          </h1>
        </div>
      </section>
    </section>
  </form>
</DocumentFragment>
`;
